src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -62,15 +67,341 @@ int
 xrtXclbinUUID(xclDeviceHandle dhdl, xuid_t out)
 {
   try {
-   api::xrtXclbinUUID(dhdl, out);
-   return 0;
+    api::xrtXclbinUUID(dhdl, out);
+    return 0;
   }
-  catch (const xrt_core::error& ex) {
+  catch (const xrt_core::error &ex) {
     xrt_core::send_exception_message(ex.what());
     return ex.get();
   }
-  catch (const std::exception& ex) {
+  catch (const std::exception &ex) {
+    xrt_core::send_exception_message(ex.what());
+    return -1;
+  }
+}
+
+namespace xrt {
+
+// class xclbin_impl - class for xclbin objects
+//
+// Life time of xclbin are managed through shared pointers.
+// A buffer is freed when last references is released.
+class xclbin_impl
+{
+protected:
+  std::vector<char> data;
+  const axlf *top;
+
+  void init_axlf_handle()
+  {
+    const axlf *tmp = (const axlf *) this->data.data();
+    this->top = strncmp(tmp->m_magic, "xclbin2", 7) ? nullptr : tmp; // Future: Do not hardcode "xclbin2"
+  }
+
+public:
+  xclbin_impl(const std::vector<char> &data)
+  {
+    this->data = data;
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  xclbin_impl(const std::string &filename)
+  {
+    if (filename.empty())
+      throw std::runtime_error("No XCLBIN specified");
+
+    // Load the file into this->data
+    std::ifstream stream(filename);
+    stream.seekg(0, stream.end);
+    size_t size = stream.tellg();
+    stream.seekg(0, stream.beg);
+    this->data.resize(size);
+    stream.read(this->data.data(), size);
+
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  virtual
+  ~xclbin_impl()
+  {
+  }
+
+  void
+  check_empty() const
+  {
+    if (this->data.size() == 0 || this->top == nullptr)
+      throw xrt_core::error(-EINVAL, "Invalid XCLBIN data");
+  }
+
+  const std::vector<std::string>
+  get_cu_names() const
+  {
+    check_empty();
+    std::vector<std::string> names;
+    std::vector<uint64_t> cus = xrt_core::xclbin::get_cus(top, false); // Why would I ever need them encoded
+    for (auto cu : cus)
+      names.push_back(xrt_core::xclbin::get_ip_name(top, cu));
+
+    return names;
+  }
+
+  const std::string
+  get_xsa_name() const
+  {
+    check_empty();
+    return reinterpret_cast<const char *>(top->m_header.m_platformVBNV);
+  }
+
+  uuid
+  get_uuid() const
+  {
+    check_empty();
+    return uuid(top->m_header.uuid);
+  }
+
+  const std::vector<char> &
+  get_data() const
+  {
+    check_empty();
+    return data;
+  }
+
+};
+
+} //namespace
+
+////////////////////////////////////////////////////////////////
+// xrt_xclbin C++ API implementations (xrt_xclbin.h)
+////////////////////////////////////////////////////////////////
+namespace xrt {
+
+xclbin::xclbin(const std::string &filename) : handle(std::make_shared<xclbin_impl>(filename))
+{}
+
+xclbin::xclbin(const std::vector<char> &data) : handle(std::make_shared<xclbin_impl>(data))
```

Comment:
Yep, if you code init_axlf_handle to throw, then the exception will propagate through this ctor all the way to user.  I think that is what we want.